### PR TITLE
Fixes in postprocessing

### PR DESCRIPTION
**Pull request by mkruckow**
This PR should fix most of open errors and warnings appearing in the post-processing.

- [x] 1) Check for `None`s before requesting dtypes from an object in the visualization [`plot1D.py`, `plot2D.py`], see issue #236 
- [x] 2) In case of start at RLO turn warning of no remaining history into a verbose output [`psygrid.py`]. When we keep track of non MT systems there this isn't worth a warning anymore.
- [x] 3) Remove warning about no RLO when cutting the histories. [`scrubbing.py`]
- [x] 4) Add a check, whether a star can collapse before trying to collapse it in binary girds with no or stable MT. [`post_processing.py`] This is needed, when we have low mass stars like in the CO-HMS grid, which stop at max_age.
- [x] 5) Adopt change of return in `do_core_collapse_BH`. [`profile_collapse.py`]
- [x] 6) Replace MT cases, if an earlier one would happen after a later one. [`termination_flags.py`, `common_functions.py`]
- [x] 7) Renaming of pipeline scripts to start with posydon. [`bin/posydon-run-pipeline`, `bin/posydon-setup-pipeline`, `pipeline_steps.rst`, `pipeline_ini.rst`, `just_step_1.ipynb`, `run_full_piepeline.ipynb`, `step_rerun.ipynb`] -> **Those changes require to reinstall posydon via** `pip install -e .`**. Additionally, old pipeline scripts need to be adopted or recreated in case to rerun them.**
- [x] 8) New ability to add unknown keys to ensure plotting [`plot2D.py`, `plot_defaults.py`], see issue #209
- [x] 9) Add cleanup jobs [`posydon-setup-pipeline`, `pipeline_ini.rst`, `pipeline_quest.ini`, `pipeline_yggdrasil.ini`]
- [x] 10) Don't use replace to detect a format specifier of a number in the title to be replaced by `all`, because it won't work in case of a value with a unit behind. Thus using a split on the equal sign instead. [`posydon-run-pipeline`]
- [x] 11) Put all checks in run-pipeline to come from flags in the csv files created in setup-pipeline [`posydon-run-pipeline`, `posydon-setup-pipeline`, `pipeline_steps.rst`, `pipeline_additions.rst`]
  - step_1
  - step_3
  - step_4
  - step_R
  - plots
- [x] 12) Add field `ORIGINAL_COMPRESSIONS` to pipeline step 3 [`posydon-setup-pipeline`, `pipeline_ini.rst`, `pipeline_quest.ini`, `pipeline_yggdrasil.ini`]
- [x] 13) Auto detect range in plotting [`posydon-run-pipeline`]
- [x] 14) Capture if date is None [`_version.py`], see issue #236 
- [x] 15) Add run directory to post-processing error to point to the dedicated issuing run [`post_processing.py`]
- [x] 16) Add option `allow_spin_None`, which is true for the post-processing models and sets the `mass`, `m_disk_accreted`, `m_disk_radiated`, and `state`, while `spin` is `None` to be ignored by the interpolator training. [`MODELS.py`, `step_SN.py`]
- [x] 17) Rephrase warning about H-rich WDs [`step_SN.py`]
- [x] 18) Don't check for `initial_z` in v1 dirnames. [`grids/io.py`]
- [x] 19) For `m_disk_accreted`, `m_disk_radiated` convert `None` to `np.nan` in collapse_star [`step_SN.py`]
- [x] 20) Enforce value types to be floats when checking interpolation errors (this avoids overflows if the dynamical type would e.g. be set to uint); ignore division by zero warnings when calculating relative errors (x/0=inf if x!=0, 0/0=nan) [`interpolation.py`]
- [x] 21) Allow to close figures after they are saved [`interpolation.py`]
- [x] 22) Add newly created directories to list of newly created files to change their group/permissions, too [`posydon-setup-pipeline`]
- [x] 23) Account for any TF2 when getting combined TF12 and expand the plotting defaults for TF12. [`combine_TF.py`, `plot_defaults.py`]
- testing
  - [x] 1-23) on CO-HMS
  - [x] 1-23) on CO-HeMS
  - [x] 1-23) on HMS-HMS
- [x] resolve conflicts introduced by other PRs merged into development, while this PR was open [`posydon-run-pipeline`, `step_mesa.py`, `step_SN.py`, `post_processing.py`, `binarypopulation.py`]
- [x] fixing bugs introduced in PR #170 [`post_processing.py`]

known issues:

- step_1:
  - `UserWarning: No RLO overflow for this binary.` (seen in CO-HeMS gird) -> done, see 3)
  - `UserWarning: Ignored MESA run because of no RLO in:` (seen in CO-HeMS gird) -> done, see 2)
- step_2 plotting:
  - ? no exception message (seen in CO-HMS gird) -> more debugging output needed / check, whether it still happens after 6) and 8) -> possibly turned to `not a PNG file`, which is strange to happen while creating a file -> not well reproducible
  - mass transfer `case_B1/A1` (seen in CO-HMS gird) -> see questions on [slack #termination_flags](https://posydon.slack.com/archives/C017XHXH2NL/p1704978227201609); done, see 6)
  - mass transfer `case_B1/C1/B1/C1` (seen in CO-HMS gird) -> see questions on [slack #termination_flags](https://posydon.slack.com/archives/C017XHXH2NL/p1704978227201609); done, see 6)
  - mass transfer `case_B1/C1/A1/C1` (seen in CO-HMS gird) -> see questions on [slack #termination_flags](https://posydon.slack.com/archives/C017XHXH2NL/p1704978227201609); done, see 6)
  - mass transfer other cases, which are unknown to the plotting defaults, but are physical -> done, see 8)
- step_3: 
  - `None` in CO type (seen in CO-HMS and CO-HeMS grid) -> see question on [slack #developers](https://posydon.slack.com/archives/CDCS3MS31/p1705049985126689?thread_ts=1702570600.013119&cid=CDCS3MS31) -> keep as is
  - `None` in SN type (seen in CO-HMS and CO-HeMS grid) -> see question on [slack #developers](https://posydon.slack.com/archives/CDCS3MS31/p1705049985126689?thread_ts=1702570600.013119&cid=CDCS3MS31) -> keep as is
  - `The star cannot collapse: star state H-rich_Core_H_burning.  TF1 max_age  interpolation class no_MT` (seen in CO-HMS grid) -> done, see 4)
  - `FAILED core collapse!  TF1 Primary got stopped before central carbon depletion  interpolation class no_MT` (seen in HMS-HMS grid) -> caused by MESA runs without final profile but considered as non-crashed, as it can happen for massive stars, where we remove some late evolution history. It would not happen if we'd allow the SN models to have `use_profiles=True` [see `MODELS.py`]. -> done, see 16)
  - `UserWarning: Invalid co/He core masses! Setting m_WD=m_star!` (seen in CO-HMS grid) -> check, whether it still happens after 4) -> still happening, warning is valid and got therefore a bit rephrased, see 17)
  - ? `Failed to print star values!` (seen in CO-HMS grid) -> more debugging output needed -> not well reproducible
  - `MODEL01 direct m_disk_accreted is not a float!` (seen in CO-HMS grid) -> done, see 5) -> still happening, more debugging -> done, see 19)
  - `MODEL01 direct m_disk_radiated is not a float!` (seen in CO-HMS grid) -> done, see 5) -> still happening, more debugging -> done, see 19)
  - ? `Donor radius from the binary class object and the profile do not agree` (seen in CO-HMS grid) -> not well reproducible
  - ? `Donor mass from the binary class object and the profile do not agree` (seen in CO-HMS grid) -> not well reproducible
  - ? `unsupported operand type(s) for ** or pow(): 'float' and 'NoneType'` (seen in CO-HMS grid) -> not well reproducible
  - ? `unsupported operand type(s) for -: 'float' and 'NoneType'` (seen in CO-HMS grid) -> not well reproducible
  - `Error during MODEL05 Patton&Sukhbold20-engineN20 core collapse prescrition!  The history did not contain core masses at He depletion!` (seen in CO-HeMS starting at RLO) -> see question on [slack #running_grids](https://posydon.slack.com/archives/C01LXP17BH7/p1708530484623489) -> see new issue #260 
- step_4:
  - ... `UserWarning: y_pred contains classes not in y_true` (seen in CO-HeMS grid) -> feedback from @prinse1545 needed -> further inspection in separate PR
  - ... `ValueError: All-NaN slice encountered` -> has a dirty workaround -> further inspection in separate PR
- step_4 plotting:
  - `RuntimeWarning: divide by zero encountered in divide` (seen in HMS-HMS and CO-HMS gird) -> done, see 20)
  - `RuntimeWarning: invalid value encountered in divide` (seen in HMS-HMS and CO-HMS gird) -> done, see 20)
  - `RuntimeWarning: More than 20 figures have been opened.` (seen in CO-HMS gird) -> done, see 21), but there might be other plotting routines not closing figures in the POSYDON code
  - `latex was not able to process the following string:  b'$m_\\\\mathrm{CO}=%1.2f\\\\,M_\\\\odot$'` (seen in CO-HMS gird) -> done, see 10)
  - ... `UserWarning: 1NN interpolation used for N binaries out of hull.` (seen in HMS-HMS and CO-HMS gird) -> feedback from @prinse1545 needed -> further inspection in separate PR
  - ... `RuntimeWarning: overflow encountered in divide` (seen in CO-HMS gird) -> done, see 20) -> still happening -> further inspection in separate PR
  - ... `KeyError: 'initial_MT'` -> check for key and add warning -> further inspection in separate PR
- ? issue #221 -> I have the impression, this is not directly caused by us (I haven't found any direct call of `is_sparse` in our code) but by other modules we are using, hence we'd probably need to rely on those to fix this.

### CODE REVIEW

**Comment by astroJeff** [comment]
I'm posting an error I see in step_1

```
Storing initial/final values and metadata to HDF5...
/projects/b1119/jja744/POSYDON/posydon/utils/gridutils.py:74: ConversionWarning: Some errors were detected !
    Line #3097 (got 1 columns instead of 51)
/projects/b1119/jja744/POSYDON/posydon/utils/gridutils.py:78: UserWarning: Problems with reading file /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/1e-02_Zsun/rerun_HeMB_MLTp_mesh_grid_low_res_combined/Zbase_0.0001_m1_15.7650_m2_9.4590_initial_z_1.4200e-04_initial_period_in_days_7.1969e+01_grid_index_539/LOGS2/history.data
/projects/b1119/jja744/POSYDON/posydon/utils/gridutils.py:74: ConversionWarning: Some errors were detected !
    Line #3097 (got 1 columns instead of 1)
```

After some investigation, it seems this arises when there is some corruption in the data files (I suspect this is what happened; the error is thrown at the last line of the history file which shows garbage, and the .out file for this run ends abruptly, without any error message not with a runtime limit). It's a rare error, only occurring for a couple systems, so I'm not sure how much time/effort we should spend trying to come up with a better error message here.

**Comment by astroJeff** [comment]
There's another warning I've found during step_2:

```
FAILED TO PLOT $q=0.85$

'case_A1/A2/A1/B1/A2'
FAILED TO PLOT $q=0.90$

'case_A1/B2/B1'
FAILED TO PLOT $q=0.99$

'case_A1/B2/B1'
```

This only happens for a couple grid slices and only at the lowest metallicities. Maybe those cases don't have any points? Again, it's a minor point, but I bet in this case it's an easy fix.

**Comment by astroJeff** [comment]
I am also seeing a bunch of errors during step_3, that look like:

```
Error during MODEL09 Sukhbold+16-engineN20 core collapse prescrition!
FAILED core collapse!
TF1 Primary got stopped before central carbon depletion
interpolation class no_MT
```

Overall, there are something like 2900 of these throughout. In each case, the error is thrown 10 times, one for each model, so a total of ~290 failed models.

**Comment by astroJeff** [comment]
For step_4, I am seeing a series of the following error messages:

```
plotting: INTERP_ERROR_age from /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/LITE/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed.h5 to /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/plots/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed
plotting: INTERP_ERROR_star_1_mass from /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/LITE/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed.h5 to /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/plots/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_proces
sed
plotting: INTERP_ERROR_star_2_mass from /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/LITE/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed.h5 to /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/plots/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_proces
sed
```

They tend to come in batches, so I think it means there is one or a few models for certain grids that cause the problem. This could be related to errors in the previous step?

**Comment by mkruckow** [comment]
> I'm posting an error I see in step_1
> 
> ```
> Storing initial/final values and metadata to HDF5...
> /projects/b1119/jja744/POSYDON/posydon/utils/gridutils.py:74: ConversionWarning: Some errors were detected !
>     Line #3097 (got 1 columns instead of 51)
> /projects/b1119/jja744/POSYDON/posydon/utils/gridutils.py:78: UserWarning: Problems with reading file /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/1e-02_Zsun/rerun_HeMB_MLTp_mesh_grid_low_res_combined/Zbase_0.0001_m1_15.7650_m2_9.4590_initial_z_1.4200e-04_initial_period_in_days_7.1969e+01_grid_index_539/LOGS2/history.data
> /projects/b1119/jja744/POSYDON/posydon/utils/gridutils.py:74: ConversionWarning: Some errors were detected !
>     Line #3097 (got 1 columns instead of 1)
> ```
> 
> After some investigation, it seems this arises when there is some corruption in the data files (I suspect this is what happened; the error is thrown at the last line of the history file which shows garbage, and the .out file for this run ends abruptly, without any error message not with a runtime limit). It's a rare error, only occurring for a couple systems, so I'm not sure how much time/effort we should spend trying to come up with a better error message here.

I'm not sure, what you expect here more, the UserWarning says, that it has trouble in reading the history.data file. That's all the information we have at this point. That the last line is corrupted we only see when checking it ourselves.

**Comment by mkruckow** [comment]
> For step_4, I am seeing a series of the following error messages:
> 
> ```
> plotting: INTERP_ERROR_age from /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/LITE/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed.h5 to /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/plots/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed
> plotting: INTERP_ERROR_star_1_mass from /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/LITE/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed.h5 to /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/plots/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_proces
> sed
> plotting: INTERP_ERROR_star_2_mass from /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/LITE/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_processed.h5 to /projects/b1119/POSYDON_GRIDS/HMS-HMS/v2/2e-01_Zsun/plots/grid_random_combined_rerun6_HeMB_MLTp_mesh_grid_proces
> sed
> ```
> 
> They tend to come in batches, so I think it means there is one or a few models for certain grids that cause the problem. This could be related to errors in the previous step?

This is not an error. It tells you that it creates the interpolation_error plots.

**Comment by astroJeff** [comment]
Update (Jan 25): We briefly discussed this PR. There are multiple steps that still need to be done - especially because this seems to include multiple issues. @mkruckow to lead effort towards tackling these bugs.

**Comment by astroJeff** [comment]
Update (Feb 15): @mkruckow is working on this. A big PR, so it will take some time.

**Comment by astroJeff** [comment]
Update (Feb 22): Almost ready - some testing on HMS-HMS grid and step_4 warnings to be resolved.

**Comment by mkruckow** [comment]
What is remaining:
- test on the HMS-HMS grid: @astroJeff or @sgossage could you do this?
- there are a few cases, which aren't well reproducible, hence I couldn't fix them and they might show up in the future again (hopefully reproducible at this time)
- there are a few things left in step_4 and the plotting of it. Here I need more details from @prinse1545 how to get there useful information, what values make the issue and where they are coming from to differentiate between a correct warning or whether there is an issue behind; about the overflow: I'm not sure whether we can ignore this because overflows usually produce wrong values, where you won't know about later, when using those values
- issue #221 looks for me to have resolved possibly by an update of the module it came from
- the enhancement of putting plots into one file with pages instead of many files, I'd be willing to postpone to later (if so, I'll open an issue for it)

All reviewers: please make a comment if there is any question/suggestion about the above!

**Comment by mkruckow** [comment]
Update (Feb 29): final test on HMS-HMS and review needed.

**Comment by maxbriel** [comment]
I tested the commit, 3b014af3cee1f32689f2c25d39e053c06044284d, on the single_HeMS grid. This seems to function correctly.

However, Matthias mentioned that PR #170 also caused issues for the post-processing of binary grids with this PR that need to be resolved.

**Comment by mkruckow** [comment]
There looks to be some more issues cause by PR #170 , which I'll inspect and try to solve.

**Comment by mkruckow** [comment]
On the HMS-HMS grid a new error appeared:
`ConversionWarning: Some errors were detected !`, this looks to be connected to some corrupted files. But most corrupted files are correctly ignored and don't cause later issues in the post-processing, I'd need to inspect what is so special about those corrupted files to cause more issues than the others.

File: posydon/grids/io.py [hunk 1]
```diff
@@ -413,7 +413,10 @@ def initial_values_from_dirname(mesa_dir):
     if "initial_mass" in dirname:                           # single-star grid
```
Comment (ZepeiX) [inline posydon/grids/io.py:413]: A minor thing here. For v1 single star grids, there is no initial_z.
if "initial_z" not in dirname:
        variable_names = ["initial_mass"]
else:
        variable_names = ["initial_mass", "initial_z"]

Comment (mkruckow) [reply posydon/grids/io.py:413]: I have added a check for v1 in the single star case

**Review by ZepeiX — DISMISSED** [review]

**Comment by maxbriel** [comment]
I'm finding a few edge cases in the interpolator around the NS vs WD boundary, where $M_\mathrm{rem}$ returns NaN.

I guess this is a population synthesis issue, not related to this PR. I will make a separate issue for this.

File: posydon/binary_evol/MESA/step_mesa.py [hunk 2]
```diff
@@ -1470,6 +1474,11 @@ def __call__(self, binary):
                                   'while the period is inside the grid.')
                 
         else:
+            if len(self.binary.state_history) > 2:
+                if self.binary.state_history[-2] == 'detached':
+                    set_binary_to_failed(self.binary)
+                    raise ValueError('CO_HMS_RLO binary outside grid and coming from detached')
+
```
Comment (maxbriel) [inline posydon/binary_evol/MESA/step_mesa.py]: These selections have been removed with a later PR on the development branch.
They should be removed for the CO_HMS_RLO and CO_HeMS_RLO grids.

CO-HeMS does not have the extra code.

```
 if len(self.binary.state_history) > 2:
                if self.binary.state_history[-2] == 'detached':
                    set_binary_to_failed(self.binary)
                    raise ValueError('CO_HMS_RLO binary outside grid and coming from detached')
```

File: posydon/binary_evol/MESA/step_mesa.py [hunk 3]
```diff
@@ -1597,6 +1606,11 @@ def __call__(self, binary):
 
         # Redirect to the detached step
         else:
+            if len(self.binary.state_history) > 2:
+                if self.binary.state_history[-2] == 'detached':
+                    set_binary_to_failed(self.binary)
+                    raise ValueError('CO_HeMS_RLO binary outside grid and coming from detached')
+
```
Comment (maxbriel) [inline posydon/binary_evol/MESA/step_mesa.py]: Same here

**Review by maxbriel — APPROVED** [review]

File: posydon/binary_evol/MESA/step_mesa.py (hunk 3 above)
Comment (mkruckow) [reply posydon/binary_evol/MESA/step_mesa.py]: removed.

File: posydon/binary_evol/MESA/step_mesa.py (hunk 2 above)
Comment (mkruckow) [reply posydon/binary_evol/MESA/step_mesa.py]: removed.

**Review by ZepeiX — APPROVED** [review]

**Pull request merged**